Source/Dafny/Compiler-cpp.cs

```diff
@@ -0,0 +1,2338 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright (C) Amazon.  All Rights Reserved.
+//
+//-----------------------------------------------------------------------------
+using System;
+using System.CodeDom;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.IO;
+using System.Diagnostics.Contracts;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using Bpl = Microsoft.Boogie;
+
+namespace Microsoft.Dafny {
+  public class CppCompiler : Compiler {
+    public CppCompiler(ErrorReporter reporter, ReadOnlyCollection<string> otherHeaders)
+    : base(reporter) {
+      this.headers = otherHeaders;
+      this.datatypeDecls = new List<DatatypeDecl>();
+      this.classDefaults = new List<string>();
+    }
+
+    private ReadOnlyCollection<string> headers;
+    private List<DatatypeDecl> datatypeDecls;
+    private List<string> classDefaults;
+    
+    /*
+     * Unlike other Dafny and Dafny's other backends, C++ cares about
+     * the order in which types are declared.  To make this more likely
+     * to succeed, we emit type information as gradually as possible
+     * in hopes that definitions are in place when needed.
```

Comment:
Would it help to emit stub declarations for types into .h files? The Java and Go back-ends have facilities to write some multiple files.

Reply:
The C++ backend does start by emitting (all of) the type declarations as stubs at the top of the single `.h` file it produces.  That helps, but unfortunately there are still some complexities for things like emitting the class definition if one class contains another (or a class contains a datatype), where C++ will object if it doesn't know the size of the type.